lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -89,6 +89,12 @@ static CodableConformanceType typeConformsToCodable(TypeChecker &tc,
   // Some generic types need to be introspected to get at their "true" Codable
   // conformance.
   auto canType = target->getCanonicalType();
+  if (auto referenceType = dyn_cast<ReferenceStorageType>(canType)) {
```

Comment:
Some other comments on other parts of the code:
- We have a better way to get out the type inside an Optional, `getAnyOptionalObjectType`
- I'm not sure encodeIfPresent is correct default behavior for optionals. An optional is still a strong reference in the object graph.